### ISSUE TITLE
Duplicate keys in block harvester and listener

### ISSUE THREAD

**Issue by Bitcoinera** [Colm3na/polkastats-backend-v3#48]
It seems the logic in these crawlers is not adjusted enough. Sometimes the crawlers try to store the same duplicated element in the DB, resulting in an ugly duplicate keys error.

The fix is actually to effectively prevent these duplicates from trying being stored, as the primary keys are correct.

Error in Block harvester:

![image](https://user-images.githubusercontent.com/40721795/76172383-7b026f80-6195-11ea-9c44-4e0d0ab4dec0.png)

Error in Block listener:

![image](https://user-images.githubusercontent.com/40721795/76172393-92d9f380-6195-11ea-9df0-14baa443d8c7.png)

Especially easy to reproduce with a slow wifi network and at startup with the new backend structure ( #46 )

**Comment by mariopino** [comment]
Good find @Bitcoinera, can you post the log with more context to help debug it? 

A slow wifi network is not our primary use case thought  ^^

**Comment by mariopino** [comment]
I'm testing https://github.com/Colm3na/polkastats-backend-v3/pull/46 dockers in a VPS and looks good, filling the database pretty fast.

**Comment by Bitcoinera** [comment]
> Good find @Bitcoinera, can you post the log with more context to help debug it?

> A slow wifi network is not our primary use case though ^^

Well, maybe they don't need to be slow networks. At least here at home it is happening every time.

These are the complete logs, 2-3 min after having launched with `node index.js`:

```
[PolkaStats backend v3] - Block listener - Detected chain reorganization at block #1367677, updating author, author name, hash and state root
[PolkaStats backend v3] - Block listener - Adding block #1367678 (0x3057…b76e)
[PolkaStats backend v3] - Block listener - => Adding event #1367678-0 system => ExtrinsicSuccess
[PolkaStats backend v3] - Block listener - => Adding event #1367678-1 system => ExtrinsicSuccess
[PolkaStats backend v3] - Block listener - => Adding event #1367678-2 system => ExtrinsicSuccess
[PolkaStats backend v3] - Starting block harvester...
[PolkaStats backend v3] - Starting active accounts crawler...
[PolkaStats backend v3] - Starting phragmen crawler...
[PolkaStats backend v3] - Block harvester - Detected gap! Harvesting blocks from #1367668 to #0
[PolkaStats backend v3] - Active Accounts - Processing account FfePrzTRhPTwQ6N8s7FqfcUB6wj2cVRDPt8LwKHPq44eDh4
[PolkaStats backend v3] - Active Accounts - Processing account FfePrzTRhPTwQ6N8s7FqfcUB6wj2cVRDPt8LwKHPq44eDh4
[PolkaStats backend v3] - Block harvester - Error adding event #1367668-1: error: duplicate key value violates unique constraint "event_pkey", sql: INSERT INTO event (
            block_number,
            event_index,
            section,
            method,
            phase,
            data
          ) VALUES (
            '1367668',
            '1',
            'system',
            'ExtrinsicSuccess',
            '{"ApplyExtrinsic":1}',
            '[{"weight":10000,"class":"Normal","paysFee":true}]'
          );
[PolkaStats backend v3] - Block harvester - Error adding event #1367668-2: error: duplicate key value violates unique constraint "event_pkey", sql: INSERT INTO event (
            block_number,
            event_index,
            section,
            method,
            phase,
            data
          ) VALUES (
            '1367668',
            '2',
            'system',
            'ExtrinsicSuccess',
            '{"ApplyExtrinsic":2}',
            '[{"weight":1000000,"class":"Normal","paysFee":true}]'
          );
[PolkaStats backend v3] - Block harvester - Error adding event #1367668-0: error: duplicate key value violates unique constraint "event_pkey", sql: INSERT INTO event (
            block_number,
            event_index,
            section,
            method,
            phase,
            data
          ) VALUES (
            '1367668',
            '0',
            'system',
            'ExtrinsicSuccess',
            '{"ApplyExtrinsic":0}',
            '[{"weight":10000,"class":"Operational","paysFee":true}]'
          );
[PolkaStats backend v3] - Active Accounts - Processing account JFJAoazqGPyi1gdYNsVDcg8fWswWmmyschqSCbdRMZfbBDw
[PolkaStats backend v3] - Active Accounts - Processing account JFJAoazqGPyi1gdYNsVDcg8fWswWmmyschqSCbdRMZfbBDw
[PolkaStats backend v3] - Active Accounts - Processing account FfePrzTRhPTwQ6N8s7FqfcUB6wj2cVRDPt8LwKHPq44dUua
[PolkaStats backend v3] - Active Accounts - Processing account FfePrzTRhPTwQ6N8s7FqfcUB6wj2cVRDPt8LwKHPq44dUua
[PolkaStats backend v3] - Active Accounts - Processing account GRLTy6VqhjysazDtLhUxCjrpxYBvQNxH9PYeptzMHZ8q3Ex
[PolkaStats backend v3] - Active Accounts - Processing account GRLTy6VqhjysazDtLhUxCjrpxYBvQNxH9PYeptzMHZ8q3Ex
[PolkaStats backend v3] - Block harvester - Error adding block #1367668: undefined
[PolkaStats backend v3] - Active Accounts - Processing account J2FqMkxGGUSQEWCLJ4zeCxySWuyhUhsfEdY9eXgBkEZsKKH
[PolkaStats backend v3] - Active Accounts - Processing account J2FqMkxGGUSQEWCLJ4zeCxySWuyhUhsfEdY9eXgBkEZsKKH
```

**Comment by mariopino** [comment]
Yes, seems to be a problem with sql query used to detect the gaps!

I added a quick fix in https://github.com/Colm3na/polkastats-backend-v3/pull/46

**Comment by Bitcoinera** [comment]
Nice! The bug is now gone, so fast!

Now there is only one TypeError showing shortly after startup:

![image](https://user-images.githubusercontent.com/40721795/76173502-5790f200-61a0-11ea-88df-2972b7e18a7d.png)

But I am sure you can kill this as fast as you did with the first one :wink:

**Comment by Bitcoinera** [comment]
Oh no! Maybe I sang victory too soon! The bug is still there... :disappointed: 

Now only a bit more delayed...

```
[PolkaStats backend v3] - Block harvester - Detected gap! Harvesting blocks from #1367666 to #0
[PolkaStats backend v3] - Active Accounts - Processing account FfePrzTRhPTwQ6N8s7FqfcUB6wj2cVRDPt8LwKHPq3zfasL
[PolkaStats backend v3] - Block harvester - Error adding event #1367665-0: error: duplicate key value violates unique constraint "event_pkey", sql: INSERT INTO event (
            block_number,
            event_index,
            section,
            method,
            phase,
            data
          ) VALUES (
            '1367665',
            '0',
            'system',
            'ExtrinsicSuccess',
            '{"ApplyExtrinsic":0}',
            '[{"weight":10000,"class":"Operational","paysFee":true}]'
          );
[PolkaStats backend v3] - Block harvester - Error adding event #1367665-1: error: duplicate key value violates unique constraint "event_pkey", sql: INSERT INTO event (
            block_number,
            event_index,
            section,
            method,
            phase,
            data
          ) VALUES (
            '1367665',
            '1',
            'system',
            'ExtrinsicSuccess',
            '{"ApplyExtrinsic":1}',
            '[{"weight":10000,"class":"Normal","paysFee":true}]'
          );
[PolkaStats backend v3] - Block harvester - Error adding event #1367665-2: error: duplicate key value violates unique constraint "event_pkey", sql: INSERT INTO event (
            block_number,
            event_index,
            section,
            method,
            phase,
            data
          ) VALUES (
            '1367665',
            '2',
            'system',
            'ExtrinsicSuccess',
            '{"ApplyExtrinsic":2}',
            '[{"weight":1000000,"class":"Normal","paysFee":true}]'
          );
```

**Comment by mariopino** [comment]
Yep. I see! it's tricky because data inserting is so fast in the harvester. When you stops the docker maybe there are rows in `block` table for block #100 but no events in `event` table for that block and vice versa (= kind of inconsistent state).

**POSSIBLE SOLUTION:** we can just insert sql deletes before the sql inserts in the first loop, so we can make sure that all data is in place in both tables (delete & insert in both tables). There is no big performance issue in rewrite a few rows in a >million rows tables to make sure the data is consistent :-)

**Comment by mariopino** [comment]
Tomorrow will be another day xDD

**Comment by mariopino** [comment]
Fixed in #46 !
